service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/aws/sagemakerNotebook/CreateAwsSagemakerNotebookStep.java

```diff
@@ -0,0 +1,85 @@
+import bio.terra.workspace.service.workspace.AwsCloudContextService;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys;
+import bio.terra.workspace.service.workspace.model.AwsCloudContext;
+import java.util.Collection;
+import java.util.HashSet;
+import software.amazon.awssdk.arns.Arn;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.services.sts.model.Tag;
+
+public class CreateAwsSagemakerNotebookStep implements Step {
+
+  private final ControlledAwsSagemakerNotebookResource resource;
+  private final AwsCloudContextService awsCloudContextService;
+  private final AuthenticatedUserRequest userRequest;
+  private final SamService samService;
+
+  public CreateAwsSagemakerNotebookStep(
+      ControlledAwsSagemakerNotebookResource resource,
+      AwsCloudContextService awsCloudContextService,
+      AuthenticatedUserRequest userRequest,
+      SamService samService) {
+    this.resource = resource;
+    this.awsCloudContextService = awsCloudContextService;
+    this.userRequest = userRequest;
+    this.samService = samService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    AwsCloudContext cloudContext =
+        awsCloudContextService.getRequiredAwsCloudContext(resource.getWorkspaceId());
+
+    AwsCredentialsProvider credentialsProvider =
+        AwsUtils.createWsmCredentialProvider(
+            awsCloudContextService.getRequiredAuthentication(),
+            awsCloudContextService.discoverEnvironment());
+
+    SamUser samUser = samService.getSamUser(userRequest);
+    Collection<Tag> tags = new HashSet<>();
+    AwsUtils.appendUserTags(tags, samUser);
+    AwsUtils.appendResourceTags(tags, cloudContext);
+
+    FlightMap inputParameters = flightContext.getInputParameters();
+    // TODO(TERRA-550): creationParameters may be used later
+    ApiAwsSagemakerNotebookCreationParameters creationParameters =
+        inputParameters.get(
+            ControlledResourceKeys.CREATE_NOTEBOOK_PARAMETERS,
+            ApiAwsSagemakerNotebookCreationParameters.class);
+
+    AwsUtils.createSageMakerNotebook(
+        credentialsProvider,
+        resource,
+        Arn.fromString(
+            inputParameters.get(
+                ControlledResourceKeys.AWS_ENVIRONMENT_USER_ROLE_ARN, String.class)),
+        Arn.fromString(
+            inputParameters.get(ControlledResourceKeys.AWS_LANDING_ZONE_KMS_KEY_ARN, String.class)),
+        Arn.fromString(
+            inputParameters.get(
+                ControlledResourceKeys.AWS_LANDING_ZONE_NOTEBOOK_LIFECYCLE_CONFIG_ARN,
+                String.class)),
+        tags);
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    // TODO(TERRA-312) add steps to stop, delete
```

Comment:
i think this is pretty important to implement in the same PR to ensure this step is idempotent. we would have orphan notebook if the flight fails and we undo. Because we would have deleted the resources in our resource Dao and not know about this notebook anymore if it's created.

Reply:
if it's small, can you just add it?

really appreciate your smaller pr but i still like it to work every pr.

Reply:
We're just talking about adding these stairway steps for undo of creation, and not plumbing the delete API, correct?

Reply:
yeah, that's what i mean, delete the cloud resources in the undo.

Reply:
Delete function added in undo - https://github.com/DataBiosphere/terra-workspace-manager/pull/1259